### ISSUE TITLE
LaTeX stops working after a certain point

### ISSUE THREAD

**Issue by theFr1nge** [mdzk-rs/mdzk#62]
When processing markdown files from obsidian, LaTeX works until a certain point in the document and then simply stops working, after that, every latex entry I make using $$ is converted to HTML as-is. Here is an example:

```markdown
## The Squeeze Theorem
When you take three funtions $f(x)$, $g(x)$ and $h(x)$ such that $f(x) \le g(x) \le h(x)$, but you can't calculate the limit of the function $g(x)$ at a value $c$, you can simply calculate $\lim_{x\to c} f(x)$ and $\lim_{x\to c} h(x)$ . The value of $\lim_{x\to c} g(x)$ must be between those values and if they are equal, then it must equal to that value as well. This method is particularly useful when dealing with the limits of periodical functions such as [[Trigonometric Functions]], for instance, when calculating $\lim_{x\to 1}$ for $g(x) = x^2sin(\frac{\pi}{x})$ . Since the range of the sine function is $-1 \le sin(x) \le 1$, we can take $f(x) = -x^2$ and $g(x) = x^2$. Now, calculating the limit of both these we functions as x approaches 0, we find that they both equal 0, therefore the answer is 0.
```

When I run mdzk build, I get this:

![image](https://user-images.githubusercontent.com/30830979/143290578-879df6f3-8264-4335-9c38-b6ce2a48bb5c.png)

The corresponding HTML generated is:

```html
<div id="content" class="content">
                    <main>
                        <h2 id="the-squeeze-theorem"><a class="header" href="#the-squeeze-theorem">The Squeeze Theorem</a></h2>
<p>When you take three funtions <span class="katex-inline"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>f</mi><mo stretchy="false">(</mo><mi>x</mi><mo stretchy="false">)</mo></mrow><annotation encoding="application/x-tex">f(x)</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height: 1em; vertical-align: -0.25em;"></span><span style="margin-right: 0.10764em;" class="mord mathnormal">f</span><span class="mopen">(</span><span class="mord mathnormal">x</span><span class="mclose">)</span></span></span></span></span>, <span class="katex-inline"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>g</mi><mo stretchy="false">(</mo><mi>x</mi><mo stretchy="false">)</mo></mrow><annotation encoding="application/x-tex">g(x)</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height: 1em; vertical-align: -0.25em;"></span><span style="margin-right: 0.03588em;" class="mord mathnormal">g</span><span class="mopen">(</span><span class="mord mathnormal">x</span><span class="mclose">)</span></span></span></span></span> and <span class="katex-inline"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>h</mi><mo stretchy="false">(</mo><mi>x</mi><mo stretchy="false">)</mo></mrow><annotation encoding="application/x-tex">h(x)</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height: 1em; vertical-align: -0.25em;"></span><span class="mord mathnormal">h</span><span class="mopen">(</span><span class="mord mathnormal">x</span><span class="mclose">)</span></span></span></span></span> such that <span class="katex-inline"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>f</mi><mo stretchy="false">(</mo><mi>x</mi><mo stretchy="false">)</mo><mo>≤</mo><mi>g</mi><mo stretchy="false">(</mo><mi>x</mi><mo stretchy="false">)</mo><mo>≤</mo><mi>h</mi><mo stretchy="false">(</mo><mi>x</mi><mo stretchy="false">)</mo></mrow><annotation encoding="application/x-tex">f(x) \le g(x) \le h(x)</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height: 1em; vertical-align: -0.25em;"></span><span style="margin-right: 0.10764em;" class="mord mathnormal">f</span><span class="mopen">(</span><span class="mord mathnormal">x</span><span class="mclose">)</span><span class="mspace" style="margin-right: 0.277778em;"></span><span class="mrel">≤</span><span class="mspace" style="margin-right: 0.277778em;"></span></span><span class="base"><span class="strut" style="height: 1em; vertical-align: -0.25em;"></span><span style="margin-right: 0.03588em;" class="mord mathnormal">g</span><span class="mopen">(</span><span class="mord mathnormal">x</span><span class="mclose">)</span><span class="mspace" style="margin-right: 0.277778em;"></span><span class="mrel">≤</span><span class="mspace" style="margin-right: 0.277778em;"></span></span><span class="base"><span class="strut" style="height: 1em; vertical-align: -0.25em;"></span><span class="mord mathnormal">h</span><span class="mopen">(</span><span class="mord mathnormal">x</span><span class="mclose">)</span></span></span></span></span>, but you can't calculate the limit of the function <span class="katex-inline"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>g</mi><mo stretchy="false">(</mo><mi>x</mi><mo stretchy="false">)</mo></mrow><annotation encoding="application/x-tex">g(x)</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height: 1em; vertical-align: -0.25em;"></span><span style="margin-right: 0.03588em;" class="mord mathnormal">g</span><span class="mopen">(</span><span class="mord mathnormal">x</span><span class="mclose">)</span></span></span></span></span> at a value <span class="katex-inline"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>c</mi></mrow><annotation encoding="application/x-tex">c</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height: 0.43056em;"></span><span class="mord mathnormal">c</span></span></span></span></span>, you can simply calculate $\lim_{x\to c} f(x)$ and <span class="katex-inline"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mrow><mi>lim</mi><mo>⁡</mo></mrow><mrow><mi>x</mi><mo>→</mo><mi>c</mi></mrow></msub><mi>h</mi><mo stretchy="false">(</mo><mi>x</mi><mo stretchy="false">)</mo></mrow><annotation encoding="application/x-tex">\lim_{x\to c} h(x)</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height: 1em; vertical-align: -0.25em;"></span><span class="mop"><span class="mop">lim</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height: 0.151392em;"><span class="" style="top: -2.55em; margin-right: 0.05em;"><span class="pstrut" style="height: 2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight">x</span><span class="mrel mtight">→</span><span class="mord mathnormal mtight">c</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height: 0.15em;"><span class=""></span></span></span></span></span></span><span class="mspace" style="margin-right: 0.166667em;"></span><span class="mord mathnormal">h</span><span class="mopen">(</span><span class="mord mathnormal">x</span><span class="mclose">)</span></span></span></span></span> . The value of $\lim_{x\to c} g(x)$ must be between those values and if they are equal, then it must equal to that value as well. This method is particularly useful when dealing with the limits of periodical functions such as <a href="Trigonometric%20Functions.html">Trigonometric Functions</a>, for instance, when calculating $\lim_{x\to 1}$ for $g(x) = x^2sin(\frac{\pi}{x})$ . Since the range of the sine function is $-1 \le sin(x) \le 1$, we can take $f(x) = -x^2$ and $g(x) = x^2$. Now, calculating the limit of both these we functions as x approaches 0, we find that they both equal 0, therefore the answer is 0.</p>
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.13.18/dist/katex.min.css" integrity="sha384-zTROYFVGOfTw7JV7KUu8udsvW2fx4lWOsCEDqhBreBwlHI4ioVRtmIvEThzJHGET" crossorigin="anonymous">
<script defer="" src="https://cdn.jsdelivr.net/npm/katex@0.13.20/dist/katex.min.js" integrity="sha384-ov99pRO2tAc0JuxTVzf63RHHeQTJ0CIawbDZFiFTzB07aqFZwEu2pz4uzqL+5OPG" crossorigin="anonymous"></script>
<script>
document.addEventListener('DOMContentLoaded', function() {
    var math = document.querySelectorAll("[class^=katex]");
    for (var i = 0; i < math.length; i++) {
      // Convert escaped ampersands before rendering
      var toRender = math[i].textContent.replace(/&amp;/g, '&');
      katex.render(toRender, math[i]);
    }
}, false)
</script>
                    </main>

                    <nav class="nav-wrapper" aria-label="Page navigation">
                        <!-- Mobile navigation buttons -->
                            <a rel="prev" href="../../../TUDelft/Q2/Calculus - CSE1200/Square Root Trick - Limit.html" class="mobile-nav-chapters previous" title="Previous chapter" aria-label="Previous chapter" aria-keyshortcuts="Left">
                                <i class="fa fa-angle-left"></i>
                            </a>

                            <a rel="next" href="../../../TUDelft/Q2/Calculus - CSE1200/Trigonometric Functions.html" class="mobile-nav-chapters next" title="Next chapter" aria-label="Next chapter" aria-keyshortcuts="Right">
                                <i class="fa fa-angle-right"></i>
                            </a>

                        <div style="clear: both"></div>
                    </nav>
                </div>
```

I can't seem to figure out what triggers this.

**Comment by kmaasrud** [comment]
I've noticed this too, and I can't wrap my head around what is causing it... It does not seem to be a consistent amount of delimiter sets either, it seems almost random how it breaks down.

A workaround while I'm trying to fix this is to split your note into more paragraphs. That helped in my case.

Thanks for submitting this bug, I'll bump it up on my priority list :smile:

**Comment by theFr1nge** [comment]
I was able to detect what is causing the bug, and came up with a potential fix.

The problem was with `pulldown_cmark`. It sometimes splits the Text event into several pieces. When the split occurs inside a LaTeX entry, the entry is split into two and it is not converted into KaTeX. The best way to solve this is by using a new function that clears up the buffer until it sees an unmatched $ sign. Here is an example, note that it does not support escaped dollar signs in this form however:

```rust
fn safeclear(buf: &mut String) {
    let mut cache = String::new();
    let mut parity = true;
    
    while let Some(c) = buf.pop(){
        cache.push(c);
    }
    buf.clear();

    while let Some(c) = cache.pop(){
        if c == '$'{
            parity = !parity;
            buf.clear();
        }
        if !parity {
            buf.push(c);
        }
    }
}
```
Another problem that causes this issue is that the Text event does not contain the links as-is. So when there is a link like 
```
[Some Text](link.md)
```
The buffer only has `Some Text`. Because of this, running `ch.content.replacen()` does not work since the buffer does not have the identical string as ch.content. One way to solve this would be passing the Chapter as a parameter to the inline function, but this is not very elegant. Is there any better approach?

**Comment by kmaasrud** [comment]
Closed in #63
